### PR TITLE
Print the list of tasks that were not executed when a build failed

### PR DESCRIPTION
**Pull request by callanoc**
**When a build fails because one of the tasks failed, some tasks following that may not get executed.
This MR includes those in the summary under “Tasks that were not executed due to failing dependencies”.**

### CODE REVIEW

File: .flake8 [hunk 1]
```diff
@@ -6,5 +6,7 @@ ignore=
     W503,
     # line break after binary operator
     W504,
+    # whitespace before ':' https://github.com/psf/black/issues/1437
+    E203
```
Comment (ntrinquier) [inline .flake8:10]: let's keep the trailing comma :)

```suggestion
    E203,
```

File: src/kraken/core/system/executor/__init__.py [hunk 2]
```diff
@@ -34,6 +34,20 @@ def set_status(self, task: Task, status: TaskStatus) -> None:
...
+        goals: bool = False,
+        pending: bool = False,
+        failed: bool = False,
+    ) -> Iterator[Task]:
+        """Returns the tasks in the graph in arbitrary order.
+
+        :param goals: Return only goal tasks (i.e. leaf nodes).
+        :param pending: Return only pending tasks.
+        :param failed: Return only failed tasks.
+        :param all: Return from all tasks, not just from the tasks that need to be executed."""
```
Comment (ntrinquier) [inline src/kraken/core/system/executor/__init__.py:49]: - the param `all` does not exist anymore
- for other params, the doc says "Return _only_ ... tasks" ➡️ this is odd: if `goals` and `pending` are both `True`, what should it return?

File: tests/test_default_executor.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    task_b = kraken_project.do("b", MyFailingTask)
+    task_c = kraken_project.do("c", MyTask)
+    task_d = kraken_project.do("d", MyTask)
+
+    task_b.add_relationship(task_a)
+    task_c.add_relationship(task_b)
+    task_d.add_relationship(task_c)
+
+    graph = TaskGraph(kraken_project.context).trim([task_d])
+    assert set(graph.tasks()) == {task_b, task_c, task_a, task_d}
```
Comment (ntrinquier) [inline tests/test_default_executor.py:83]: not a big deal, but we might as well sort them alphabetically

```suggestion
    assert set(graph.tasks()) == {task_a, task_b, task_c, task_d}
```

File: src/kraken/core/system/executor/__init__.py (hunk 2 above)
Comment (callanoc) [reply src/kraken/core/system/executor/__init__.py:49]: It seems like it's returning tasks according to the last True boolean in order goals then pending then failed https://github.com/kraken-build/kraken-core/blob/develop/src/kraken/core/system/graph.py#L312

File: .flake8 (hunk 1 above)
Comment (callanoc) [reply .flake8]: Fixed

File: src/kraken/core/system/executor/__init__.py (hunk 2 above)
Comment (callanoc) [reply src/kraken/core/system/executor/__init__.py]: Modified

File: src/kraken/core/system/executor/default.py [hunk 4]
```diff
@@ -151,6 +153,16 @@ def after_execute_graph(self, graph: Graph) -> None:
...
                     self.format_duration(f"[{self._duration[task_path]:.3f}s]") if task_path in self._duration else "",
                 )
+                if status.is_failed():
+                    failed_tasks.append(task)
+
+        if len(failed_tasks) != 0:
+            print(flush=True)
+            print(self.format_header("Tasks that were not executed due to failing dependencies"), flush=True)
+            print(flush=True)
+            for task in graph.tasks(goals=False, pending=True):  # Keeping only the pending tasks
```
[the remark this answers is not in the document]
Comment (callanoc) [reply src/kraken/core/system/executor/default.py]: Modified

File: tests/test_default_executor.py (hunk 3 above)
Comment (callanoc) [reply tests/test_default_executor.py]: Fixed

File: tests/test_default_executor.py [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+    task_c.add_relationship(task_b)
+    task_d.add_relationship(task_b)
+
+    graph = TaskGraph(kraken_project.context).trim([group])
+    execute_print_test(graph)
+    captured = capsys.readouterr()
+    assert skipped_tests_title in captured.out
+    result = trim_printed_result(captured)
+    assert ":c" in result
+    assert ":d" in result
```
[the remark this answers is not in the document]
Comment (callanoc) [reply tests/test_default_executor.py]: Updated

File: tests/test_default_executor.py [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+
+
+def test__DefaultExecutor__print_correct_failures_with_independent_groups(kraken_project: Project, capsys: Any) -> None:
+    """This test tests if when a task failed within one group,
+     successor tasks with dependencies will be printed as failed.
+    Even if another group is following, there is no impact on it
```
[the remark this answers is not in the document]
Comment (callanoc) [reply tests/test_default_executor.py]: Updated

File: src/kraken/core/system/executor/default.py [hunk 7]
```diff
@@ -151,6 +152,14 @@ def after_execute_graph(self, graph: Graph) -> None:
...
                     self.format_duration(f"[{self._duration[task_path]:.3f}s]") if task_path in self._duration else "",
                 )
+        not_executed_tasks = list(graph.tasks(not_executed=True))
+        if len(not_executed_tasks) != 0:
+            print(flush=True)
+            print(self.format_header("Tasks that were not executed due to failing dependencies"), flush=True)
+            print(flush=True)
+            for task in not_executed_tasks:  # Keeping only the pending tasks
+                print(" " * (len(self.execute_prefix) + 1) + task.path)  # path seems to be
+                # Deprecated in v0.13.0 (task.py l231)
```
Comment (ntrinquier) [inline src/kraken/core/system/executor/default.py:162]: - `# Keeping only the pending tasks` should no longer be here
  - it should be above, but in fact I'd argue it should not exist at all, it is just paraphrasing the function call `not_executed_tasks = list(graph.tasks(not_executed=True))` :)
- `path` is indeed deprecated, but IMO the comment is not adding any value

```suggestion
            for task in not_executed_tasks:
                print(" " * (len(self.execute_prefix) + 1) + task.path)
```

File: src/kraken/core/system/graph.py [hunk 8]
```diff
@@ -294,13 +294,14 @@ def tasks(
...
         failed: bool = False,
+        not_executed: bool = False,
     ) -> Iterator[Task]:
         """Returns the tasks in the graph in arbitrary order.
 
         :param goals: Return only goal tasks (i.e. leaf nodes).
         :param pending: Return only pending tasks.
         :param failed: Return only failed tasks.
-        :param all: Return from all tasks, not just from the tasks that need to be executed."""
+        :param not_executed: Return not executed tasks (failed dependency)"""
```
Comment (ntrinquier) [inline src/kraken/core/system/graph.py:304]:
```suggestion
        :param not_executed: Return only not executed tasks (i.e. downstream of failed tasks)"""
```

**Review by cdbrkfxrpt — APPROVED** [review]

**Pull request merged**